libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/SyncTask.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2019-present, salesforce.com, inc.
+ * All rights reserved.
+ * Redistribution and use of this software in source and binary forms, with or
+ * without modification, are permitted provided that the following conditions
+ * are met:
+ * - Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * - Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * - Neither the name of salesforce.com, inc. nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission of salesforce.com, inc.
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.salesforce.androidsdk.smartsync.manager;
+
+
+import com.salesforce.androidsdk.analytics.EventBuilderHelper;
+import com.salesforce.androidsdk.rest.RestClient;
+import com.salesforce.androidsdk.smartstore.store.SmartStore;
+import com.salesforce.androidsdk.smartsync.util.SmartSyncLogger;
+import com.salesforce.androidsdk.smartsync.util.SyncState;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+/**
+ * Abstract super class of runnable classes responsible for running syncs
+ */
+public abstract class SyncTask implements Runnable {
```

Comment:
We don't have to change this now, but in the future we might want to consider using `AsyncTask` here instead of `Runnable`. `AsyncTask` is more Android lifecycle friendly and it's easier to get callbacks and make UI changes if required on the main thread directly.

Reply:
Looking at https://developer.android.com/topic/performance/threads.html, it says:
> For this reason, we suggest that you only use AsyncTask to handle work items shorter than 5ms in duration.
So not quite our use case.

I also considered using java.concurrent classes like FutureTask. However, cancellation relies on Thread.interrupt() (see https://github.com/frohoff/jdk8u-jdk/blob/master/src/share/classes/java/util/concurrent/FutureTask.java#L174) but that doesn't play nice with OkHttp (tons of issues around it if you google for it).

Reply:
Cool! 👍